File: flutter/lib/src/log4agent.dart
import 'dart:convert';
import 'dart:math';

import 'package:dio/dio.dart';
import 'package:http/http.dart' as http;

import 'log4agent_level.dart';
import 'log4agent_config.dart';
import 'log4agent_redactor.dart';
import 'log4agent_transport.dart';

class Log4Agent {
  Log4Agent._();

  static Log4AgentConfig _config = const Log4AgentConfig();
  static Log4AgentTransport _transport = Log4AgentHttpTransport();
  static String _sessionId = _newSessionId();
  static bool _sessionStarted = false;
  static bool _initialized = false;

  static void configure({
    Log4AgentConfig config = const Log4AgentConfig(),
    http.Client? client,
    Dio? dio,
    Log4AgentTransport? transport,
    bool force = false,
  }) {
    if (_initialized && !force) return;
    _initialized = true;
    _config = config;
    _sessionId = _newSessionId();
    _sessionStarted = false;
    _transport.close();
    _transport =
        transport ??
        (dio != null
            ? Log4AgentDioTransport(dio)
            : Log4AgentHttpTransport(client));
    _startSession();
  }

  static void debug(
    String category,
    String message, {
    Map<String, String> attributes = const {},
  }) {
    log(category, message, level: Log4AgentLevel.debug, attributes: attributes);
  }

  static void info(
    String category,
    String message, {
    Map<String, String> attributes = const {},
  }) {
    log(category, message, level: Log4AgentLevel.info, attributes: attributes);
  }

  static void warn(
    String category,
    String message, {
    Map<String, String> attributes = const {},
  }) {
    log(category, message, level: Log4AgentLevel.warn, attributes: attributes);
  }

  static void error(
    String category,
    String message, {
    Map<String, String> attributes = const {},
  }) {
    log(category, message, level: Log4AgentLevel.error, attributes: attributes);
  }

  static void log(
    String category,
    String message, {
    Log4AgentLevel level = Log4AgentLevel.info,
    Map<String, String> attributes = const {},
  }) {
    final current = _config;
    if (!current.enabled || current.endpoint.isEmpty) return;
    _startSession();
    final safeMessage = current.redactionEnabled
        ? Log4AgentRedactor.redactUrl(message, keys: current.redactionKeys)
        : message;
    final safeAttributes = current.redactionEnabled
        ? Log4AgentRedactor.redactAttributes(
            attributes,
            keys: current.redactionKeys,
          )
        : attributes;
    final payload = jsonEncode({
      'timestamp': DateTime.now().toUtc().toIso8601String(),
      'app': current.app,
      'deviceId': current.deviceId,
      'sessionId': _sessionId,
      'platform': 'flutter',
      'level': level.name,
      'category': category,
      'message': safeMessage,
      'attributes': safeAttributes,
    });
    _transport.postJson(current.endpoint, payload).ignore();
  }

  static void _startSession() {
    final current = _config;
    if (_sessionStarted || !current.enabled || current.endpoint.isEmpty) return;
    _sessionStarted = true;
    _transport
        .postJson(
          _sessionEndpoint(current.endpoint),
          jsonEncode({
            'timestamp': DateTime.now().toUtc().toIso8601String(),
            'app': current.app,
            'deviceId': current.deviceId,
            'sessionId': _sessionId,
            'platform': 'flutter',
          }),
        )
        .ignore();
  }

  static String _sessionEndpoint(String logEndpoint) {
    if (logEndpoint.endsWith('/logs')) {
      return '${logEndpoint.substring(0, logEndpoint.length - '/logs'.length)}/sessions';
    }
    return '${logEndpoint.replaceFirst(RegExp(r'/+$'), '')}/sessions';
  }

  static String _newSessionId() {
    final now = DateTime.now().toUtc().millisecondsSinceEpoch;
    final random = Random().nextInt(0x7fffffff).toRadixString(16);
    return '$now-$random';
  }
}

File: kmp/log4agent/src/commonMain/kotlin/dev/log4agent/Log4Agent.kt
package dev.log4agent

import io.ktor.client.HttpClient
import io.ktor.client.plugins.HttpTimeout
import io.ktor.client.request.headers
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlin.random.Random
import kotlin.time.Clock
import kotlin.time.ExperimentalTime

object Log4Agent {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val json = Json { encodeDefaults = false }

    private var config: Log4AgentConfig = Log4AgentConfig()
    private var client: HttpClient = newClient(config)
    private var ownsClient: Boolean = true
    private var sessionId: String = newSessionId()
    private var sessionStarted: Boolean = false
    private var initialized: Boolean = false

    fun configure(config: Log4AgentConfig, force: Boolean = false) {
        configure(config = config, httpClient = null, force = force)
    }

    fun configure(config: Log4AgentConfig, httpClient: HttpClient?, force: Boolean = false) {
        if (initialized && !force) return
        initialized = true
        this.config = config
        sessionId = newSessionId()
        sessionStarted = false
        if (httpClient != null) {
            closeOwnedClient()
            client = httpClient
            ownsClient = false
            startSession()
            return
        }
        closeOwnedClient()
        client = newClient(config)
        ownsClient = true
        startSession()
    }

    fun configure(
        endpoint: String = Log4AgentEndpoint.default(),
        enabled: Boolean = true,
        redactionEnabled: Boolean = true,
        httpClient: HttpClient? = null,
        force: Boolean = false,
    ) {
        configure(
            config = Log4AgentConfig(
                enabled = enabled,
                endpoint = endpoint,
                redactionEnabled = redactionEnabled,
            ),
            httpClient = httpClient,
            force = force,
        )
    }

    fun setEnabled(enabled: Boolean) {
        config = config.copy(enabled = enabled)
    }

    fun debug(category: String, message: String, attributes: Map<String, String> = emptyMap()) =
        log(category, message, Log4AgentLevel.Debug, attributes)

    fun info(category: String, message: String, attributes: Map<String, String> = emptyMap()) =
        log(category, message, Log4AgentLevel.Info, attributes)

    fun warn(category: String, message: String, attributes: Map<String, String> = emptyMap()) =
        log(category, message, Log4AgentLevel.Warn, attributes)

    fun error(category: String, message: String, attributes: Map<String, String> = emptyMap()) =
        log(category, message, Log4AgentLevel.Error, attributes)

    fun log(
        category: String,
        message: String,
        level: Log4AgentLevel = Log4AgentLevel.Info,
        attributes: Map<String, String> = emptyMap(),
    ) {
        log(Log4AgentEvent(category = category, message = message, level = level, attributes = attributes))
    }

    @OptIn(ExperimentalTime::class)
    fun log(event: Log4AgentEvent) {
        val current = config
        if (!current.enabled || current.endpoint.isBlank()) return
        startSession()

        val message = if (current.redactionEnabled) {
            Log4AgentRedactor.redactMessage(event.message, current.redactionKeys)
        } else {
            event.message
        }
        val attributes = if (current.redactionEnabled) {
            Log4AgentRedactor.redactAttributes(event.attributes, current.redactionKeys)
        } else {
            event.attributes
        }

        val payload = buildJsonObject {
            put("timestamp", Clock.System.now().toString())
            put("app", current.app)
            put("deviceId", current.deviceId.ifBlank { log4AgentPlatformName() })
            put("sessionId", sessionId)
            put("platform", log4AgentPlatformName())
            put("level", event.level.name.lowercase())
            put("category", event.category)
            put("message", message)
            put("attributes", JsonObject(attributes.mapValues { JsonPrimitive(it.value) }))
        }

        reportLog4AgentQueued(event.category, current.endpoint)
        scope.launch {
            runCatching {
                val response = client.post(current.endpoint) {
                    headers {
                        append(HttpHeaders.ContentType, ContentType.Application.Json.toString())
                    }
                    setBody(json.encodeToString(JsonObject.serializer(), payload))
                }
                reportLog4AgentResult(event.category, response.status.value)
            }.onFailure { error ->
                reportLog4AgentFailure(error.message ?: error::class.simpleName ?: "unknown")
            }
        }
    }

    @OptIn(ExperimentalTime::class)
    private fun startSession() {
        val current = config
        if (sessionStarted || !current.enabled || current.endpoint.isBlank()) return
        sessionStarted = true
        val payload = buildJsonObject {
            put("timestamp", Clock.System.now().toString())
            put("app", current.app)
            put("deviceId", current.deviceId.ifBlank { log4AgentPlatformName() })
            put("sessionId", sessionId)
            put("platform", log4AgentPlatformName())
        }
        scope.launch {
            runCatching {
                client.post(sessionEndpoint(current.endpoint)) {
                    headers {
                        append(HttpHeaders.ContentType, ContentType.Application.Json.toString())
                    }
                    setBody(json.encodeToString(JsonObject.serializer(), payload))
                }
            }.onFailure { error ->
                reportLog4AgentFailure(error.message ?: error::class.simpleName ?: "unknown")
            }
        }
    }

    private fun sessionEndpoint(logEndpoint: String): String =
        if (logEndpoint.endsWith("/logs")) logEndpoint.removeSuffix("/logs") + "/sessions"
        else logEndpoint.trimEnd('/') + "/sessions"

    private fun newSessionId(): String =
        "${Clock.System.now().toEpochMilliseconds()}-${Random.nextInt().toUInt().toString(16)}"

    private fun newClient(config: Log4AgentConfig): HttpClient =
        HttpClient(defaultLog4AgentEngineFactory()) {
            expectSuccess = false
            install(HttpTimeout) {
                connectTimeoutMillis = config.connectTimeoutMillis
                requestTimeoutMillis = config.requestTimeoutMillis
                socketTimeoutMillis = config.socketTimeoutMillis
            }
            applyLog4AgentPlatformConfig()
        }

    private fun closeOwnedClient() {
        if (ownsClient) client.close()
    }
}
